### PR TITLE
cps/new way of applying dirichlet

### PR DESCRIPTION
**Pull request by inigolcanalejo**
In this PR we introduce a new way of applying the dirichlet bcs. It makes it flexible for any domain shape and for any angle of attack.

This was the idea and work of @marcnunezc.

Attention: this process will be migrated to c++ soon!

### CODE REVIEW

File: applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py
Comment (philbucher) [inline applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py:5]: same comment as in the other PR

File: applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py
Comment (philbucher) [inline applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py]:
```suggestion

```
this is legacy, not needed in new code

File: applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py
Comment (philbucher) [inline applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py]: why does every condition need this?

Comment (inigolcanalejo) [reply applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py]: https://github.com/KratosMultiphysics/Kratos/pull/4661#discussion_r275655518

File: applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py
Comment (inigolcanalejo) [reply applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py:5]: Do you mean this comment https://github.com/KratosMultiphysics/Kratos/pull/4642#discussion_r274904021? I understand and agree that it is better to do this kind of thing in c++, however, we should find a solution until we completely migrate this process to c++. Ideas:

1. We delete this function and use numpy (the idea of @marcnunezc originally was to avoid to import numpy to reduce the dependency on external libraries as commented in other reviews)
2. We delete this function and do it by hand.

Any other idea?

File: applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py
Comment (msandre) [inline applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py]: The code until this line will be easier to understand if you package it in its own function with a name like `FindFarthestUpstreamBoundaryNode()`

File: applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py
Comment (msandre) [inline applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py:83]: The remaining code should go into a second function like `AssignFarFieldBoundaryCondition()`. Then this function becomes easier to understand / change:

```
def Execute(self):
    x0, y0, z0 = self.FindFarthestUpstreamBoundaryPosition()
    self.AssignFarFieldBoundaryCondition(x0, y0, z0)
```

File: applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py
Comment (msandre) [inline applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py]: Should `self.velocity_infinity` be a unit vector? If so, please check this in the constructor. Also, consider the choice of name for easier understanding. E.g., `free_stream_direction`.

**Review by msandre — CHANGES_REQUESTED** [review]
I didn't repeat comments regarding naming from #4668. I suggest merging #4668 first and then fixing conflicts in this PR. In the meantime you can check my comments specific to this PR. If you split up the functions as suggested, you can also test the code more easily. I would suggest adding some small python unit tests for both functions.

File: applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py
Comment (inigolcanalejo) [reply applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py]: self.velocity_infinity is a vector equal to the magnitude to the velocity of the undisturbed free stream far away from the body. It is therefore not a unit vector. I changed the name to free_stream_velocity as you suggested in https://github.com/KratosMultiphysics/Kratos/pull/4668 to make it more clear.

File: applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py
Comment (inigolcanalejo) [reply applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py:83]: I spitted the code according to your review and I tried to make it more clear. Now the class has the following functions:
```python
_FindFarthestUpstreamBoundaryNode()
_AssignFarFieldBoundaryConditions()
_AssignDirichletFarFieldBoundaryCondition
_AssignNeumannFarFieldBoundaryCondition
```

I also added some comments and renamed some variables, the implementation looks now like this:
```python3
    def Execute(self):
        reference_inlet_node = self._FindFarthestUpstreamBoundaryNode()
        self._AssignFarFieldBoundaryConditions(reference_inlet_node)

    def _FindFarthestUpstreamBoundaryNode(self):
        # The farthes upstream boundary node is the node with smallest
        # projection of its distante to a given reference node onto
        # the free stream velocity.

        # Select a random node in the mesh as reference (e.g. the first node)
        for node in self.far_field_model_part.Nodes:
            reference_node = node
            break

        # Find the farthest upstream boundary node
        temporal_smallest_projection = 1e30
        for node in self.far_field_model_part.Nodes:
            distance_to_reference = node - reference_node
            # Projecting the distance to the reference onto the free stream velocity
            distance_projection = DotProduct(distance_to_reference, self.free_stream_velocity)

            if(distance_projection < temporal_smallest_projection):
                temporal_smallest_projection = distance_projection
                reference_inlet_node = node

        return reference_inlet_node

    def _AssignFarFieldBoundaryConditions(self, reference_inlet_node):
        # A Dirichlet condition is applyed at the inlet nodes and
        # a Neumann condition is applyed at the outlet nodes

        for cond in self.far_field_model_part.Conditions:
            normal = cond.GetValue(KratosMultiphysics.NORMAL)

            # Computing the projection of the free stream velocity onto the normal
            velocity_projection = DotProduct(normal, self.free_stream_velocity)

            if( velocity_projection < 0):
                # A negative projection means inflow (i.e. inlet condition)
                self._AssignDirichletFarFieldBoundaryCondition(reference_inlet_node, cond)
            else:
                # A positive projection means outlow (i.e. outlet condition)
                self._AssignNeumannFarFieldBoundaryCondition(cond)

    def _AssignDirichletFarFieldBoundaryCondition(self, reference_inlet_node, cond):
        for node in cond.GetNodes():
            # Computing the value of the potential at the inlet
            inlet_potential = DotProduct( node - reference_inlet_node, self.free_stream_velocity)

            # Fixing the potential at the inlet nodes
            node.Fix(CPFApp.VELOCITY_POTENTIAL)
            node.SetSolutionStepValue(CPFApp.VELOCITY_POTENTIAL,0,inlet_potential + self.inlet_potential_0)

            # Applying Dirichlet condition in the adjoint problem
            if self.far_field_model_part.HasNodalSolutionStepVariable(CPFApp.ADJOINT_VELOCITY_POTENTIAL):
                node.Fix(CPFApp.ADJOINT_VELOCITY_POTENTIAL)
                node.SetSolutionStepValue(CPFApp.ADJOINT_VELOCITY_POTENTIAL,0,inlet_potential)

    def _AssignNeumannFarFieldBoundaryCondition(self, cond):
        cond.SetValue(CPFApp.VELOCITY_INFINITY, self.free_stream_velocity)
```

File: applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py
Comment (msandre) [inline applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py]: I believe the `reference_node` has no effect on finding the `reference_inlet_node`. So you can simplify this code further.

File: applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py
Comment (msandre) [inline applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py]: typos

**Review by msandre — CHANGES_REQUESTED** [review]

File: applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py
Comment (inigolcanalejo) [reply applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py]: You are completely right, I do not know why we are doing this. One could directly do:

```python
def _FindFarthestUpstreamBoundaryNode(self):
        # The farthes upstream boundary node is the node with smallest
        # projection of its position vector onto the free stream velocity.

        # Find the farthest upstream boundary node
        temporal_smallest_projection = 1e30
        for node in self.far_field_model_part.Nodes:
            # Projecting the node position vector onto the free stream velocity
            distance_projection = DotProduct(node, self.free_stream_velocity)

            if(distance_projection < temporal_smallest_projection):
                temporal_smallest_projection = distance_projection
                reference_inlet_node = node

        return reference_inlet_node
```

File: applications/CompressiblePotentialFlowApplication/python_scripts/compute_lift_jump_process_2d.py
Comment (rubenzorrilla) [inline applications/CompressiblePotentialFlowApplication/python_scripts/compute_lift_jump_process_2d.py]:
```suggestion
        self.free_stream_velocity = settings["free_stream_velocity"].GetVector()
```
I think that this should work.

**Review by msandre — DISMISSED** [review]

**Comment by msandre** [comment]
Approving from my side. Please merge #4668 first and consider @rubenzorrilla's suggestion before merging this.

**Comment by RiccardoRossi** [comment]
why this is not merged yet?

File: applications/CompressiblePotentialFlowApplication/python_scripts/compute_lift_jump_process_2d.py
Comment (inigolcanalejo) [reply applications/CompressiblePotentialFlowApplication/python_scripts/compute_lift_jump_process_2d.py]: This works! I included your suggestion in https://github.com/KratosMultiphysics/Kratos/pull/4668. Since compute_lift_jump_process_2d.py is anyway removed in https://github.com/KratosMultiphysics/Kratos/pull/4668, I will leave it here like this.

**Comment by inigolcanalejo** [comment]
> why this is not merged yet?

I am waiting to merge https://github.com/KratosMultiphysics/Kratos/pull/4668 first.

**Review by rubenzorrilla — DISMISSED** [review]
Approving so you can merge as soon as #4668 is merged.

**Comment by inigolcanalejo** [comment]
This is ready to be merged.

Short question: @rubenzorrilla after you approved I had to still merge with master and solve some conflicts. When I commited the merge I automatically dismissed your approval. Is there a way around this or is this how it is supposed to be? I was just wondering because people keep approving PRs and I keep dismissing their approvals as soon as I merge with master.

**Comment by rubenzorrilla** [comment]
> This is ready to be merged.
> 
> Short question: @rubenzorrilla after you approved I had to still merge with master and solve some conflicts. When I commited the merge I automatically dismissed your approval. Is there a way around this or is this how it is supposed to be? I was just wondering because people keep approving PRs and I keep dismissing their approvals as soon as I merge with master.

This is how it is supposed to be. In any case I think it's good, once the changes have been reviewed it's straightforward to slight review and approve it again.

**Review by rubenzorrilla — APPROVED** [review]

**Comment by inigolcanalejo** [comment]
Ok, thanks for the info! :+1:

**Pull request merged**
